### PR TITLE
Feature/logging messagetemplate compat

### PR DESCRIPTION
**Pull request by tkrafael**
Continuing https://github.com/dotnet/runtime/pull/79038:

- Added documentation to private static methods, so it makes easier future contributions;
- Added unit testing for https://github.com/dotnet/runtime/issues/78013
- Renaming methods so they're semantically correct when working with '$' symbols in variables
- using Span<char> when possible
- Small bug during variable validation in parser: (src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:358 in this commit)

### CODE REVIEW

**Comment by tarekgh** [comment]
@tkrafael can you please resolve the conflict?

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Emitter.cs [hunk 1]
```diff
@@ -59,8 +59,9 @@ private static bool UseLoggerMessageDefine(LoggerMethod lm)
                     // make sure the order of the templates matches the order of the logging method parameter
                     for (int i = 0; i < lm.TemplateList.Count; i++)
                     {
-                        string t = lm.TemplateList[i];
-                        var (template, parameter) = SanitizeAtSign(t, lm.TemplateParameters[i].CodeName);
+                        var template = SanitizeSpecialSign(lm.TemplateList[i].AsSpan());
+                        var parameter = SanitizeSpecialSign(lm.TemplateParameters[i].CodeName.AsSpan());
```
Comment (layomia) [inline src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Emitter.cs:63]: Please don't use `var` - https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md.

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Emitter.cs [hunk 2]
```diff
@@ -606,16 +608,28 @@ private static string ConvertEndOfLineAndQuotationCharactersToEscapeForm(string
...
-        private static (string template, string parameter) SanitizeAtSign(string template, string parameter)
-        {
-            static string SanitizeSingle(string input) => input.Length > 0 && input[0] == '@' ? input.Substring(1) : input;
-            return (SanitizeSingle(template), SanitizeSingle(parameter));
-        }
+        /// <summary>
+        /// Checks if string contains a special symbol ('@') as starting char
+        /// </summary>
+        /// <param name="input">true if contains special symbol, otherwise false.</param>
+        /// <returns></returns>
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Emitter.cs:615]: > <returns></returns> [](http://example.com/codeflow?start=11&length=20)

please fill this one

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Emitter.cs [hunk 3]
```diff
@@ -606,16 +608,28 @@ private static string ConvertEndOfLineAndQuotationCharactersToEscapeForm(string
...
+        private static bool ContainsSpecialSymbol(ReadOnlySpan<char> input)
+            => input.Length > 0 && input[0] == '@';
+
+        /// <summary>
+        /// Removes '@' from input, allowing local variables to be declared when creating code.
+        /// </summary>
+        /// <param name="input">input that might contain '@' symbol</param>
+        /// <returns>If input starts with a symbol ('@'), returns a new string starting with underscore with '@' symbol removed, otherwise returns current string.</returns>
+        private static ReadOnlySpan<char> ProtectSpecialSymbolInVariable(ReadOnlySpan<char> input) =>
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Emitter.cs:624]: >ProtectSpecialSymbolInVariable [](http://example.com/codeflow?start=42&length=30)

I suggest renaming this method to something like `NormalizeSymbolName` or something similar. The term `Protect` is not clear here.

Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Emitter.cs:624]: >input [](http://example.com/codeflow?start=92&length=5)

You may change this as symboleName or variableName. or something similar.

Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Emitter.cs:624]: >ReadOnlySpan<char> [](http://example.com/codeflow?start=23&length=18)

passing and returning string will be better here. The callsite will be more cleaner and using spans is not helping in anything here.

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Emitter.cs [hunk 4]
```diff
@@ -83,7 +84,7 @@ namespace {lc.Namespace}
                 }
 
                 LoggerClass parent = lc.ParentClass;
-                var parentClasses = new List<string>();
+                List<string> parentClasses = new List<string>();
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Emitter.cs:87]: >List<string>  [](http://example.com/codeflow?start=16&length=13)

Sorry, it is ok to use `var` if you are doing `new`. The whole point is the code reader want to know the type. With `new` it is already clear. but for other cases it is not and that is why we avoid using `var` there.

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Emitter.cs [hunk 5]
```diff
@@ -606,16 +608,30 @@ private static string ConvertEndOfLineAndQuotationCharactersToEscapeForm(string
...
-
-        private static (string template, string parameter) SanitizeAtSign(string template, string parameter)
-        {
-            static string SanitizeSingle(string input) => input.Length > 0 && input[0] == '@' ? input.Substring(1) : input;
-            return (SanitizeSingle(template), SanitizeSingle(parameter));
-        }
+        /// <summary>
+        /// Checks if variableOrTemplateName contains a special symbol ('@') as starting char
+        /// </summary>
+        /// <param name="variableOrTemplateName"></param>
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Emitter.cs:614]: >>< [](http://example.com/codeflow?start=48&length=2)

please fill this part.

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Emitter.cs [hunk 6]
```diff
@@ -606,16 +608,30 @@ private static string ConvertEndOfLineAndQuotationCharactersToEscapeForm(string
...
+        /// <returns>true if contains special symbol, otherwise false.</returns>
```
[the remark this answers is not in the document]
Comment (tkrafael) [reply src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Emitter.cs:615]: Sorry, english is not my mother language.

[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Emitter.cs]: > english is not my mother language.

Me neither :-) we are in the same boat.

**Comment by tarekgh** [comment]
The failures do not look related to the changes. It shows in other PRs like https://github.com/dotnet/runtime/pull/80392. Let's wait for the CI to finish and we can try restart these failures.

**Review by tarekgh — APPROVED** [review]

**Pull request merged**